lib/charms/layer/openstack.py

```diff
@@ -32,52 +32,50 @@ def get_credentials():
     """
     config = hookenv.config()
 
+    creds_data = {}
+
+    # try to use Juju's trust feature
+    try:
+        log('Checking credentials-get for credentials')
+        result = subprocess.run(['credential-get'],
+                                check=True,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE)
+        _creds_data = yaml.load(result.stdout.decode('utf8'))
+        creds_data.update(_normalize_creds(_creds_data))
+    except FileNotFoundError:
+        pass  # juju trust not available
+    except subprocess.CalledProcessError as e:
+        if 'permission denied' not in e.stderr.decode('utf8'):
+            raise
+
     # try credentials config
     if config['credentials']:
         try:
-            creds_data = b64decode(config['credentials']).decode('utf8')
-            creds_data = json.loads(creds_data)
             log('Using "credentials" config values for credentials')
-            _save_creds(creds_data)
-            return True
+            _creds_data = b64decode(config['credentials']).decode('utf8')
+            creds_data.update(_normalize_creds(json.loads(_creds_data)))
         except Exception:
             status.blocked('invalid value for credentials config')
             return False
-    no_creds_msg = 'missing credentials; set credentials config'
 
     # try individual config
-    if any([config['auth-url'],
+    creds_data.update(_normalize_creds(config))
+
+    if all([config['auth-url'],
             config['username'],
             config['password'],
             config['project-name'],
             config['user-domain-name'],
             config['project-domain-name'],
             config['region']]):
-        log('Using individual config values for credentials')
-        _save_creds(config)
-        return True
-
-    # try to use Juju's trust feature
-    try:
-        result = subprocess.run(['credential-get'],
-                                check=True,
-                                stdout=subprocess.PIPE,
-                                stderr=subprocess.PIPE)
-        creds_data = yaml.load(result.stdout.decode('utf8'))
-
-        log('Using credentials-get for credentials')
         _save_creds(creds_data)
```

Comment:
This isn't what you want here, is it? just saving what was set before?

Reply:
Wait, I see now. This is saving creds_data, which is updated above, but the if is checking config, which seems a little odd.

Reply:
You are correct.